tests/Doctrine/Tests/DBAL/Functional/MasterSlaveConnectionTest.php

```diff
@@ -139,4 +143,15 @@ public function testMasterSlaveConnectionCloseAndReconnect()
         $conn->connect('master');
         $this->assertTrue($conn->isConnectedToMaster());
     }
+
+    public function testMasterSlaveFailSlave()
+    {
+        $params = $this->prepareConnectionParameters();
+        $invalidParams = $params['master'];
+        $invalidParams['password'] = 'MUUHAAAAHAAAA';
```

Comment:
I think this test exposes a problem with your implementation (and the test should be adapted too): the only exception to be caught should be a connection failed exception (due to timeout or similar), and not an authentication or setup failure.

Having the entire thing crash on mis-configuration is actually intentional, whereas the functionality you really want here is retrying on failed connection, rather than authentication.